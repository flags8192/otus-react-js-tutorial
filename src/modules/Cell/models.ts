export enum ICellInfo {
  alive = 1,
  dead = 0,
}

export interface CellArray {
  size: number;
  data: ICellInfo[];
}

export type ISelectedCells = {
  [index: number]: boolean;
};
