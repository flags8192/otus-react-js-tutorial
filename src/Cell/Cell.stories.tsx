import React, { useState, FC } from 'react';
import { ComponentStory, ComponentMeta } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { Cell } from './Cell';
import { withKnobs, number, boolean } from '@storybook/addon-knobs';

export default {
  component: Cell,
  decorators: [withKnobs],
  title: 'Example/Cell',
} as ComponentMeta<typeof Cell>;

const Template: ComponentStory<typeof Cell> = (args) => <Cell {...args} />;

export const Static = Template.bind({});

Static.args = {
  num: number('num', 3),
  alive: boolean('alive', true),
  onClick: action('clicked'),
  isRight: boolean('isRight', true),
  isBottom: boolean('isBottom', true),
  isLeft: boolean('isLeft', true),
};

export const Dynamic: FC = () => {
  const [alive, setAlive] = useState(true);
  const num = number('num', 3);
  const onClick = () => {
    setAlive(!alive);
  };

  return (
    <Cell
      num={num}
      alive={alive}
      onClick={onClick}
      isRight={boolean('isRight', true)}
      isBottom={boolean('isBottom', true)}
      isLeft={boolean('isLeft', true)}
    />
  );
};
