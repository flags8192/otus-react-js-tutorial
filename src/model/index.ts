import { IState } from '@src/model/IState';
import { IAction } from '@src/model/IAction';

export { IAction, IState };
