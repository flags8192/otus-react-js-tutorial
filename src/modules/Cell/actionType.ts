const prefix = 'CELL__';

export default {
  SET_SELECTED: `${prefix}SET_SELECTED`,
};
