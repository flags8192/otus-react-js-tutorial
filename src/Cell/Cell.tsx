import React from 'react';
import styled from '@emotion/styled';
import {CELL_HEIGHT, CELL_WIDTH} from '@src/Cell/constatnts';

interface IContainer {
    isRight?: boolean;
    isLeft?: boolean;
    isBottom?: boolean;
    frameColor?: string;
    alive?: boolean;
}

const Container = styled.div<IContainer>`
  display: block;
  width: ${({isRight}) => CELL_WIDTH - (isRight ? 2 : 1)}px;
  height: ${({isBottom}) => CELL_HEIGHT - (isBottom ? 2 : 1)}px;
  text-align: center;
  border-top: ${({frameColor}) => '1px solid ' + frameColor};
  border-left: ${({frameColor}) => '1px solid ' + frameColor};
  border-right: ${({isRight, frameColor}) => '1px solid ' + isRight ? frameColor : ''}
  border-bottom: ${({isBottom, frameColor}) => '1px solid ' + isBottom ? frameColor : ''}
  cursor: pointer;
  float: left;
  ${({isLeft}) => (isLeft ? 'clear: both;' : '')}
  transition-property: background;
  transition-duration: 0.3s;
  transition-timing-function: ease-out;
  background: #555;
  background: ${({alive}) => (alive ? '#e44' : '#555')};
`;

interface IContent {
    alive?: boolean;
}

const Content = styled.span<IContent>`
  background: ${({alive}) => (alive ? '#e44' : '#555')};
`;

export interface CellProps {
    num: number;
    alive?: boolean;
    onClick: (num: number) => void;
    isLeft?: boolean;
    isRight?: boolean;
    isBottom?: boolean;
}

export const Cell: React.FC<CellProps> = ({
                                              num,
                                              alive = true,
                                              isLeft = true,
                                              isRight = true,
                                              isBottom = true,
                                              ...rest
                                          }) => {
    const handleClick = () => {
        rest.onClick(num);
    };
    const frameColor = '#ddd';

    return (
        <Container
            role="cellContainer"
            onClick={handleClick}
            isRight={isRight}
            isLeft={isLeft}
            alive={alive}
            frameColor={frameColor}
        >
            <Content role="cellContent" alive={alive}/>
        </Container>
    );
};
