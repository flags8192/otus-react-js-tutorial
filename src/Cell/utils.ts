import { CellInfo } from '@src/Cell/models';

export const getInverted = (cell: CellInfo): CellInfo =>
  cell === CellInfo.alive ? CellInfo.dead : CellInfo.alive;
