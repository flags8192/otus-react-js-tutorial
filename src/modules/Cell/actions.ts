import actionType from './actionType';

const setSelectedCell = (cellId: number, selected: boolean) => ({
  type: actionType.SET_SELECTED,
  payload: { cellId, selected },
});

export { setSelectedCell };
