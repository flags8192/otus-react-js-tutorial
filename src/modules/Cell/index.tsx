import React, { useContext } from 'react';
import styled from '@emotion/styled';
import { CELL_SIZE } from '@src/modules/Cell/constatnts';
import store from '@src/store';
import { setSelectedCell } from '@src/modules/Cell/actions';
import ICellInfo from '@src/model/ICellInfo';

interface IContainer {
  isRight?: boolean;
  isLeft?: boolean;
  isBottom?: boolean;
  frameColor?: string;
}
const Container = styled.div<IContainer>`
  display: block;
  width: ${({ isRight }) => CELL_SIZE - (isRight ? 2 : 1)}px;
  height: ${({ isBottom }) => CELL_SIZE - (isBottom ? 2 : 1)}px;
  text-align: center;
  border-top: ${({ frameColor }) => `1px solid ${frameColor}`};
  border-left: ${({ frameColor }) => `1px solid ${frameColor}`};
  border-right: ${({ isRight, frameColor }) => `1px solid ${isRight ? frameColor : ''};`}
  border-bottom: ${({ isBottom, frameColor }) => `1px solid ${isBottom ? frameColor : ''};`}
  float: left;
  ${({ isLeft }) => (isLeft ? 'clear: both;' : '')}
  transition-property: background;
  transition-duration: 0.3s;
  transition-timing-function: ease-out;
  background: #555;
  background: #d1d1d1;
  border: 0.6vh solid transparent;
  border-color: white #9e9e9e #9e9e9e white;  
`;

interface IContent {
  selected?: boolean;
}
const Content = styled.div<IContent>`
  cursor: pointer;
  background: #ffff00;
  border-radius: 50%;
  background: ${({ selected }) => (selected ? '#ffff00' : '#ffff0070')};
  border-radius: 50%;
  width: 30px;
  height: 30px;
  margin: 5px;
  border: 0.6vh solid transparent;
  border-color: ${({ selected }) => (selected ? '#555' : '#9e9e9e')};

  :hover {
    background: #ffff00;
  }
`;

export interface CellProps {
  num: number;
  isFilled?: ICellInfo;
  isSelected?: boolean;
  isLeft?: boolean;
  isRight?: boolean;
  isBottom?: boolean;
}

const Cell: React.FC<CellProps> = ({
  num,
  isFilled = 0,
  isSelected = false,
  isLeft = true,
  isRight = true,
}) => {
  const { dispatch } = useContext(store);

  const handleClick = () => {
    dispatch(setSelectedCell(num, !isSelected));
  };
  const frameColor = '#ddd';

  return (
    <Container
      role={`cellContainer-${num}`}
      isRight={isRight}
      isLeft={isLeft}
      frameColor={frameColor}
    >
      {!!isFilled && (
        <Content role={`cellContent-${num}`} selected={isSelected} onClick={handleClick} />
      )}
    </Container>
  );
};

export default Cell;
