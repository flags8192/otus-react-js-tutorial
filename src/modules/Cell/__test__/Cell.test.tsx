import React from 'react';
import Cell from '@src/modules/Cell';
import { render, screen, fireEvent } from '@testing-library/react';
import Store from '@src/store';
import { IAction, IState } from '@src/model';

let dispatch: (action: IAction) => void;
let state: IState;

describe('Cell', () => {
  beforeEach(() => {
    dispatch = jest.fn();
    state = {
      gameFieldSize: 3,
      gameFieldData: [1, 1, 1, 1, 1, 1, 1, 1, 1],
      gameLevel: '1',
      gameFieldPercentFilled: 100,
      selectedCells: {},
    };
  });

  it('renders <Cell> using renderer cell with value=2', () => {
    const { asFragment } = render(
      <Store.Provider value={{ dispatch, state }}>
        <Cell num={2} isFilled={1} isSelected={true} />
      </Store.Provider>
    );
    expect(asFragment()).toMatchSnapshot();
    const cellContainer = screen.getByRole(/cellContainer/gi);
    expect(cellContainer).toBeInTheDocument();
    const cellContent = screen.getByRole(/cellContent-2/gi);
    expect(cellContent).toBeInTheDocument();
  });

  it('Click on cell', () => {
    render(
      <Store.Provider value={{ dispatch, state }}>
        <Cell num={2} isFilled={1} isSelected={true} />
      </Store.Provider>
    );
    const cellContainer = screen.getByRole(/cellContent-2/i);
    fireEvent.click(cellContainer);
    expect(dispatch).toHaveBeenCalledTimes(1);
    expect(cellContainer).toHaveStyle('background: #ffff00');
  });
});
