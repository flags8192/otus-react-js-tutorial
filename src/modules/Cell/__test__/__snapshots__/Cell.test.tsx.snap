// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell renders <Cell> using renderer cell with value=2 1`] = `
<DocumentFragment>
  <div
    class="css-upvruk"
    role="cellContainer-2"
  >
    <div
      class="css-214etd"
      role="cellContent-2"
    />
  </div>
</DocumentFragment>
`;
